sql: small view descriptor clean-ups

knz opened the pull request:
These are common functionality from #15388 which are of general interest, before I complete/refactor the remainder of the patch in #15388 into a different PR.

knz committed:
sql: simplify view queries during inspection

If a view is defined by a query over table `d.t`, and this view's
query is inspected (using either SHOW CREATE VIEW or pg_catalog) in
the context of database `d`, ensure that the prefix `d.` is hidden.

This is useful to ease reuse of the view definition when migrating
databases.

knz committed:
sql: avoid dropping dependencies of view descriptors twice

This hidden bug cropped up while investigating different ways to
encode view descriptors.

knz committed:
sql: support numeric table references for views

This patch makes it possible to use the numeric table reference syntax
with a selection of columns (e.g. `[123(1,2,3)]`) when the reference
is for a view. This selects a subset of the view's definition.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17475)
<!-- Reviewable:end -->

knz committed:
sql: improve the definition of views with different column names than the query

Prior to this patch, a view defined with `CREATE VIEW v(x) FROM SELECT
y FROM ...` would only store the name "x" in the descriptor. This
would cause queries to pg_catalog or information_schema to hide the
fact the view renders a column named "x" (only showing "SELECT y
...").

This patch improves upon this situation by forcing an explicit rename
in the view query if the view definition changes the names.

This was the most glaring limitation of the previous approach, but in
addition to this a bug was lurking: a view created with `CREATE VIEW
v(rowid) FROM TABLE t`, which *intends* to reveal the `rowid` such
that subsequently `SELECT * FROM v` actually selects `rowid`, didn't
work as intended: when the view plan was expanded, the `rowid` column
would be picked from the underlying plan, which would make it hidden,
which would subsequently cause `SELECT * FROM v` to not pick any
column at all. The explicit rename approach taken in this patch fixes
this bug. This bug probably never affected anyone given how unlikely
the use case is, however fixing this fits the general effort of view
descriptor improvements this patch is part of.

knz commented:
Note that even if we push the work to fix #10023/#10083 to 1.2, I think this PR should still make it to 1.1.

knz commented:
All right, this is rebased and again gtg. PTAL
